domain/modelagent/modelmigration/import.go

```diff
@@ -0,0 +1,231 @@
+// part of a model.
+type importUnitAgentBinaryOperation struct {
+	baseAgentBinaryImportOperation
+}
+
+// Execute will attempt to set the reported tools/agentbinary version for each
+// machine in the model. This operation assumes that prior to this being
+// performed all of the machines in the model have been imported into this
+// controller.
+//
+// The following errors can be expected:
+// - [github.com/juju/juju/core/errors.NotValid] if the reportedVersion is
+// not valid or the machine name is not valid.
+// - [github.com/juju/juju/core/errors.NotSupported] if the architecture is
+// not supported.
+// - [github.com/juju/juju/domain/machine/errors.MachineNotFound] when the
+// machine does not exist.
+// - [github.com/juju/juju/domain/machine/errors.MachineDead] when the
+// machine is dead.
+func (i *importMachineAgentBinaryOperation) Execute(
+	ctx context.Context,
+	model description.Model,
+) error {
+	for _, machine := range model.Machines() {
+		tools := machine.Tools()
+
+		mName := coremachine.Name(machine.Id())
+		binVer, err := semversion.ParseBinary(tools.Version())
+		if err != nil {
+			return errors.Errorf(
+				"parsing agent version for machine %q during migration: %w",
+				mName.String(), err,
+			)
+		}
+		agentBinaryVersion := coreagentbinary.Version{
+			Number: binVer.Number,
+			Arch:   binVer.Arch,
+		}
+
+		err = i.importService.SetMachineReportedAgentVersion(
+			ctx, mName, agentBinaryVersion,
+		)
+
+		if err != nil {
+			return errors.Errorf(
+				"setting reported agent version for machine %q during migration: %w",
+				mName.String(), err,
+			)
+		}
+	}
+
+	return nil
+}
+
+// Execute will attempt to set the reported tools/agentbinary version for each
+// unit of each application in the model. This operation assumes that priort to
+// this being performed all of the applications and units in the model have been
+// imported into this controller.
+//
+// The following errors are possible:
+// - [github.com/juju/juju/core/errors.NotValid] when the reportedVersion
+// is not valid.
+// - [github.com/juju/juju/core/errors.NotSupported] when the architecture
+// is not supported.
+// - [github.com/juju/juju/domain/application/errors.UnitNotFound] when the
+// unit does not exist.
+// - [github.com/juju/juju/domain/application/errors.UnitIsDead] when the
+// unit is dead.
+func (i *importUnitAgentBinaryOperation) Execute(
+	ctx context.Context,
+	model description.Model,
+) error {
+	for _, application := range model.Applications() {
+		for _, unit := range application.Units() {
+			tools := unit.Tools()
+
+			uName := coreunit.Name(unit.Name())
+			binVer, err := semversion.ParseBinary(tools.Version())
+			if err != nil {
+				return errors.Errorf(
+					"parsing agent version for unit %q during migration: %w",
+					uName.String(), err,
+				)
+			}
+			agentBinaryVersion := coreagentbinary.Version{
+				Number: binVer.Number,
+				Arch:   binVer.Arch,
```

Comment:
Same here.
If we don't use it, we don't need to export them.

Reply:
Yeah this is tricky. We know that 3.6 is going to send us the encoded bin format where 4.0 could export a new format. However this is just a string value in the description package and so no meaning can be applied to it between description versions.

We essentially have keep with what was so we don't break. I have put as part of the clean up task for this epic to look at seeing if we can modify description to remove this unwanted mess but it would require us teaching description about the value that is contained within this string.

I think what we will do is leave version as a string and introduce a new field call arch. That way we can safely say that arch is seperate and then we can treat version as just a number.